From 9f7931eadf4c9071d3b273f2e622b5263d60a4dd Mon Sep 17 00:00:00 2001
From: Philippe Gerum <rpm@xenomai.org>
Date: Thu, 12 Jun 2025 17:56:10 +0200
Subject: [PATCH 0949/1013] evl/net: dev: sanitize error path on failure to
 enable oob port

The original code had serious issues on failure to enable an oob port:
bad return status, missing thread cleanup (rx) and so on. In addition,
a failure to enable the oob mode in the device driver was not even
received and propagated downstream.

Apply the required fixes.

Signed-off-by: Philippe Gerum <rpm@xenomai.org>
---
 kernel/evl/net/dev.c | 26 +++++++++++++++++---------
 1 file changed, 17 insertions(+), 9 deletions(-)

diff --git a/kernel/evl/net/dev.c b/kernel/evl/net/dev.c
index bb1f8c764c52..37a56106b3f6 100644
--- a/kernel/evl/net/dev.c
+++ b/kernel/evl/net/dev.c
@@ -184,8 +184,10 @@ static int enable_oob_port(struct net_device *dev,
 		evl_init_flag(&est->tx_flag);
 		kt = start_handler_thread(real_dev, evl_net_do_tx,
 					KTHREAD_TX_PRIO, "tx");
-		if (IS_ERR(kt))
+		if (IS_ERR(kt)) {
+			ret = PTR_ERR(kt);
 			goto fail_start_tx;
+		}
 
 		est->tx_handler = kt;
 	}
@@ -194,7 +196,9 @@ static int enable_oob_port(struct net_device *dev,
 		evl_init_crossing(&rnds->crossing);
 
 	/* Divert traffic from the real device. */
-	netif_enable_oob_diversion(real_dev);
+	ret = netif_enable_oob_diversion(real_dev);
+	if (ret)
+		goto fail_enable_diversion;
 queue:
 	netif_enable_oob_port(dev);
 
@@ -204,18 +208,20 @@ static int enable_oob_port(struct net_device *dev,
 
 	return 0;
 
+fail_enable_diversion:
+	if (est->tx_handler) {
+		evl_stop_kthread(est->tx_handler);
+		evl_destroy_flag(&est->tx_flag);
+	}
 fail_start_tx:
 	/*
 	 * No skb has flowed yet, no pending recycling op. Likewise,
 	 * we cannot have any rxq in the cache or dump lists.
 	 */
-	if (netdev_is_oob_capable(real_dev)) {
-		evl_stop_kthread(est->rx_handler);
-		evl_destroy_flag(&est->tx_flag);
-	}
+	evl_stop_kthread(est->rx_handler);
+	evl_destroy_flag(&est->rx_flag);
 fail_start_rx:
 	evl_net_dev_purge_pool(real_dev);
-	evl_destroy_flag(&est->rx_flag);
 fail_build_pool:
 	evl_net_free_qdisc(est->qdisc);
 fail_alloc_qdisc:
@@ -285,12 +291,14 @@ static void disable_oob_port(struct net_device *dev) /* inband, rtnl_lock held *
 	netif_disable_oob_diversion(real_dev);
 
 	evl_stop_kthread(est->rx_handler);
-	if (est->tx_handler)
+	evl_destroy_flag(&est->rx_flag);
+	if (est->tx_handler) {
 		evl_stop_kthread(est->tx_handler);
+		evl_destroy_flag(&est->tx_flag);
+	}
 
 	__set_rx_filter(est, NULL);
 	evl_net_dev_purge_pool(real_dev);
-	evl_destroy_flag(&est->rx_flag);
 	/* Once the TX kthread is stopped, qdisc is idle. */
 	evl_net_free_qdisc(est->qdisc);
 	kfree(est);
-- 
2.30.2

